pkg/binary/binary.go

```diff
@@ -0,0 +1,145 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package binary
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//nolint
+const (
+	// GOOS labels
+	LINUX  = "linux"
+	DARWIN = "darwin"
+	WIN    = "windows"
+
+	// GOARCH Architecture labels
+	I386    = "386"
+	AMD64   = "amd64"
+	ARM     = "arm"
+	ARM64   = "arm64"
+	PPC     = "ppc"
+	PPC64LE = "ppc64le"
+	S390    = "s390"
+	RISCV   = "riscv"
+)
+
+// Binary is the base type of the package. It abstracts a binary executable
+type Binary struct {
+	options *Options
+	binaryImplementation
+}
+
+// Options to control the binary checker
+type Options struct {
+}
+
+// DefaultOptions set of options
+var DefaultOptions = &Options{}
+
+// New creates a new binary instance.
+func New(filePath string) (bin *Binary, err error) {
+	// Get the right implementation for the specified file
+	return NewWithOptions(filePath, DefaultOptions)
+}
+
+// NewWithOptions creates a new binary with the specified options
+func NewWithOptions(filePath string, opts *Options) (bin *Binary, err error) {
+	// Get the right implementation for the specified file
+	impl, err := getArchImplementation(filePath, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting arch implementation")
+	}
+	return &Binary{
+		options:              opts,
+		binaryImplementation: impl.(binaryImplementation),
+	}, nil
+}
+
+// getArchImplementation returns the implementation that corresponds
+// to the specified binary
+func getArchImplementation(filePath string, opts *Options) (impl interface{}, err error) {
+	// First, if file has a .exe extension, we test it first
+	// because most likely it is a win executable
+	if strings.HasSuffix(filePath, ".exe") {
+		impl, err = NewPEBinary(filePath, opts)
+		if err != nil {
+			return nil, errors.Wrap(err, "checking .exe file")
+		}
+		if impl != nil {
+			return impl, nil
+		}
+	}
+
+	// Check if we're dealing with a Linux binary
+	elf, err := NewELFBinary(filePath, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "while checking if file is an ELF binary")
+	}
+	if elf != nil {
+		return elf, nil
+	}
+
+	// Check if its a darwin binary
+	impl, err = NewMachOBinary(filePath, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "checking if file is a Mach-O binary")
+	}
+	if impl != nil {
+		return impl, nil
+	}
+
+	// Finally we check to see if it's a windows binary
+	impl, err = NewPEBinary(filePath, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "checking .exe file")
+	}
+	if impl != nil {
+		return impl, nil
+	}
```

Comment:
What is the reason behind checking twice for windows binaries?

Reply:
I figured that if a file has an .exe extension, then we're 99.999% certain that it is a win binary. So since I knew the extension already I thought I could skip the Linux & Mac scanning and save a bit on I/O. But its only like 100 bytes, I think I'll drop it.